### PR TITLE
Sphinx-gallery into nilearn 

### PR DESCRIPTION
**Pull request by Titan-C**
This pull request uses the today published sphinx-gallery-0.0.5 in pypi to generate the gallery.
It also updates some examples to a python3 compatible syntax, so they can be executed within sphinx-gallery. As sphinx-gallery already imports from future print and division

I'm listening to comments and bugs.
the old gen_rst.py shall be deleted before merge.

### CODE REVIEW

**Comment by Titan-C** [comment]
The commit SHA: f9700219589a05026b8e56d0950fb3f0a311f743
requires the sphinx-gallery 0.0.6
It is now capable to resolve links for used modules in the examples.

**Comment by lesteve** [comment]
@Titan-C you have a tiny merge conflict in one of the examples. Would you mind rebasing on master and fixing it?

**Comment by lesteve** [comment]
Also there is a new example examples/manipulating/visualizing/plot_atlas.py that needs parentheses around its print statement.

**Comment by Titan-C** [comment]
Done

**Comment by lesteve** [comment]
Thanks I think you should remove the old gallery generating code, for example gen_rst.py and related css.

Also I think what @GaelVaroquaux had in mind was shipping a version of sphinx-gallery in nilearn/doc/sphinxext rather than requiring to do `pip install sphinx-gallery`.

**Comment by GaelVaroquaux** [comment]
> Also I think what @GaelVaroquaux had in mind was shipping a version of
> sphinx-gallery in nilearn/doc/sphinxext rather than requiring to do pip
> install sphinx-gallery.

Yes.

**Comment by Titan-C** [comment]
I was also wondering about that. If one keeps a local version of the sphinx-gallery extension, and probably every project will keep its own local copy. What was the aim of keeping the centralized extension? wont all versions end up branching again? Isn't it better to use the git version installed in `develop` mode more practical?
I was having a look at the scikit-learn externals. It is a static set of code that gets updated manually, why? and why the extra install scripts within that module?

**Comment by lesteve** [comment]
> I was also wondering about that. If one keeps a local version of the sphinx-gallery extension, and probably every project will keep its own local copy. What was the aim of keeping the centralized extension? wont all versions end up branching again? Isn't it better to use the git version installed in develop mode more practical?

I guess it's up to each project to make a choice whether they want to ship sphinx-gallery or not. The main point of doing so is pure convenience for people generating the doc.

About sphinx-gallery evolving separately in different projects, there is an implicit agreement that only released versions of sphinx-gallery should be used and that potential changes should be propagated upstream.

> I was having a look at the scikit-learn externals. It is a static set of code that gets updated manually, why? and why the extra install scripts within that module?

I guess convenience again is the main reason. Users are spared having to install a few packages in order to use scikit-learn. I think the technical term is vendorizing, you may want to see what the internet has to say about it. I didn't know about the extra install scripts in external so I can't help on this one.

**Comment by GaelVaroquaux** [comment]
> If one keeps a local version of the sphinx-gallery extension, and
> probably every project will keep its own local copy. What was the aim
> of keeping the centralized extension? wont all versions end up
> branching again?

Keeping a local version doesn't mean that we will branch. The idea is to
have a synchronization script that copies exactly the sphinx-gallery code
in the projects that use it. In these projects it should never be
modified. One example of this pattern is how joblib is integrated in
scikit-learn, in sklearn/externals.

> Isn't it better to use the git version installed in develop mode more
> practical?

More practical for you. It raises two problems. One is of technicality.
It's one more thing to learn and master for a contributor. These things
pile up and make it harder and harder to contribute to a project. The
second problem is that it means that any backword incompatible change or
bug introduce in sphinx-gallery will break the projects using it.
Decoupling is a good thing.

**Comment by lesteve** [comment]
Something I didn't spot right away. We added some additional text at the top of the examples gallery and this was added directly in gen_rst.py

nilearn_examples_header

For now a work-around would be to move this text to examples/README.rst.

**Comment by lesteve** [comment]
Great job, this looks pretty close to me! I am regenerating the doc and I'll have a closer look later this afternoon.

**Comment by lesteve** [comment]
Tiny comment: I am wondering whether the joblib import in doc_resolv.py should not try to import it from sklearn.externals as a fall-back if joblib is not installed.

The good: no extra package to install to generate the doc for nilearn, scikit-learn, and other packages users that have scikit-learn installed.

The not that great: a tiny bit of scikit-learn specific code for projects that are potentially unrelated. 

I was thinking something along those lines (but with a better error message):
```python
try:
    import joblib
except ImportError as exc_joblib:
    try:
        from sklearn.externals import joblib
    except ImportError as exc_sklearn_joblib:
        exc_sklearn_joblib.args += ('joblib or scikit-learn needs to be installed',)
        raise
```

**Comment by GaelVaroquaux** [comment]
> Tiny comment: I am wondering whether the joblib import in doc_resolv.py should
> not try to import it from sklearn.externals as a fall-back if joblib is not
> installed.

:+1:. We cannot have a joblib dependency.

I would actually favor not using joblib but a simple shelve, and a test
to see if the URL is already in the shelve.

Here is some pseudo code to implement what I have in mind:
<pre>
import shelve
mem = shelve.open('foo')
if not url in mem:
    data = get_url(url)
    mem[url] = data
    mem.sync()
else:
    data = mem[url]
</pre>

**Comment by lesteve** [comment]
I don't know the code very well so it's hard for me to estimate how easy the shelve solution is and whether it would take some time to get right.

I would be inclined to say let's just use the joblib + sklearn.externals.joblib fall-back for now so that we can merge this PR.

As far as I know, the two projects that are closer to start using sphinx-gallery are nilearn and scikit-learn anyway.

**Comment by Titan-C** [comment]
> Tiny comment: I am wondering whether the joblib import in doc_resolv.py should not try to import it from sklearn.externals as a fall-back if joblib is not installed.

Joblib was originally called from scikit-learn externals. That is how is was implemented in scikit-learn. But in the early stages of development that was a pain with the continuous integration in travis as I had to additionally put scipy and scikit-learn as dependencies and installing joblib is way more convenient. So although users of sphinx-gallery are for now, users of scikit-learn, and almost everyone tends to install scipy in their home computers. When dealing with online deployment(travis, readthedocs), it becomes a pain to have to put scikit-learn as dependency. I would not incorporate sklearn.externals.joblib back, that would mean I also have to test upon it.

>I don't know the code very well so it's hard for me to estimate how easy the shelve solution is and whether it would take some time to get right.
I also don't know well the shelve solution, I'll take me some more time to learn it.

**Comment by GaelVaroquaux** [comment]
> I would be inclined to say let's just use the joblib +
> sklearn.externals.joblib fall-back for now so that we can merge this
> PR.

I'd rather not. Creeping dependencies are a problem, and we must tackle
them. The pseudo code that I have given demonstrates how easy it is to
use shelve.

> As far as I know, the two projects that are closer to start using
> sphinx-gallery are nilearn and scikit-learn anyway.

Well, as Oscar is mentionning, it's already posing problems for CI.

But anyhow, let us foster adoption by having little dependencies.

**Comment by lesteve** [comment]
> Well, as Oscar is mentionning, it's already posing problems for CI. 

Fair point, let's go for the shelve way then.

**Comment by lesteve** [comment]
>> Well, as Oscar is mentionning, it's already posing problems for CI. 

> Fair point, let's go for the shelve way then.

Actually not such a fair point, since the import joblib would work for the sphinx-gallery CI but I agree the shelve solution doesn't seem so hard and is the right way to do it.

**Comment by Titan-C** [comment]
This includes the shelve and is rebased to the current master. There are more examples than in the current nilearn website. In my computer there seems to be 2 examples that don't get executed. I can't figure out why? On the rest it does work.

**Comment by lesteve** [comment]
Thanks a lot for rebasing on master! The nilearn website hasn't been updated since the latest release so that would explain why it doesn't feature examples that have been added recently.

Ideally what we want is to make a sphinx-gallery release and use it in nilearn. Do you think this is feasible?

File: examples/decoding/plot_simulated_data.py [hunk 1]
```diff
@@ -41,9 +41,9 @@ def create_simulation_data(snr=0, n_samples=2 * 100, size=12, random_state=1):
     w[-roi_size:, -roi_size:, 0:roi_size] = 0.5
```
Comment (lesteve) [inline examples/decoding/plot_simulated_data.py]: I think this file should be kept unchanged because this changes are not really related to this PR or are they?

Comment (Titan-C) [inline examples/decoding/plot_simulated_data.py]: This is the only remaining part of the py2 py3 compatibility, as all print functions disapeared on rebase. This is needed because sphinx-gallery imports from future division, so it really requires integers as keys it crashes otherwise.

**Comment by Titan-C** [comment]
> Ideally what we want is to make a sphinx-gallery release and use it in nilearn. Do you think this is feasible?

Publishing sphinx-gallery as 0.0.7 with the shelve is straight forward, can upload to pypi now. But I wouldn't launch it yet as a stable release. As some to do from @GaelVaroquaux are still missing. https://github.com/sphinx-gallery/sphinx-gallery/issues/20
Moreover I want to test a bit longer on the configuration dictionary key naming, test the nametuple as well. Since after a stable release one shall not change the user interphase.

**Comment by lesteve** [comment]
> As some to do from @GaelVaroquaux are still missing. sphinx-gallery/sphinx-gallery#20

They could be tackled in a further release.

> Moreover I want to test a bit longer on the configuration dictionary key naming, test the nametuple as well. Since after a stable release one shall not change the user interphase.

Fair enough, how long do you need roughly to be reasonably confident there is no major issues?

Because sphinx-gallery is not used massively at the moment backward-incompatible changes are not such a big deal I would say.

File: examples/decoding/plot_simulated_data.py (hunk 1 above)
Comment (lesteve) [inline examples/decoding/plot_simulated_data.py]: I can run this example in master with python 3 so I don't think this change is absolutely needed.

Using float as numpy array indices is deprecated indeed as the following snippet shows so feel free to open a separate PR with this change

```python

import warnings

import numpy as np

warnings.simplefilter('always', DeprecationWarning)

arr = np.arange(10)

print(arr[2.3:5.9])
```

**Comment by lesteve** [comment]
@Titan-C any news on this front? We may do a release of nilearn soonish, it'd be great to start using sphinx-gallery!

I think the main things we need is:
* a sphinx-gallery release with the shelve functionality (maybe it has already happened I haven't checked)
* a script to update the nilearn sphinx-gallery copy similary to what is done for joblib in scikit-learn. TBH, it would be fine to have a simple copy as a first step and work on this script in a separate PR.

**Comment by Titan-C** [comment]
Yes, few days ago I updated to version 0.0.7 which includes the shelve and the dictionary configuration. Then the quick release of 0.0.8 is because I get a bug with CSS as it conflicts with some of the Sphinx themes. The update script is already there and it worked for the last 2 updates.
- You can run this branch. There are some examples that in my computer don't work, as I claimed before.

**Comment by Titan-C** [comment]
> You can run this branch. There are some examples that in my computer don't work, as I claimed before.

- I merged locally master into this to test. All examples work!
- But now one has to rename all images in the documentation, as now they are numbered with 3 digits (001). So the carousel it the star page can't find the image as neither other places in the documentation

**Comment by lesteve** [comment]
> Yes, few days ago I updated to version 0.0.7 which includes the shelve and the dictionary configuration. Then the quick release of 0.0.8 is because I get a bug with CSS as it conflicts with some of the Sphinx themes. The update script is already there and it worked for the last 2 updates.

Great stuff, I'll take a closer look this afternoon!

 Don't hesitate to add a comment when you push commits into your PR branch and you think things are in a good shape. This way we get notifications and it's easier to get a feeling what is going on with the project.

**Comment by lesteve** [comment]
@Titan-C, I did some quick and dirty renaming of the pngs in https://github.com/lesteve/nilearn/commit/e00ca5006febccd9928b22b4ea04af862b73e1f9. You should be able to cherry-pick it easily into your PR branch.

I did a few sanity checks and it seems fine but it'd be great if you could double-check too. A good comparison would be the documentation that Gaël generated for one of his course: http://gaelvaroquaux.github.io since nilearn.github.io is trailing a lot behind master atm.

**Comment by Titan-C** [comment]
It doesn't work in my home computer. Certainly this needs a new issue. But I can't get 2 examples to work at home. 
- plot_probabilistic_atlas_extraction.py
- plot_inverse_covariance_connectome.py
https://gist.github.com/Titan-C/7eab460917461adc61a4
there seems like the data has changes, but has not been updated. How to do it?

Apart from this, and everything that links to this images, I don't see much any difference.

**Comment by lesteve** [comment]
Hmmm works for me, try `rm ~/nilearn_data/msdl_atlas/ -rf` and rerunning the examples which are failing.

**Comment by Titan-C** [comment]
OK it works ! all examples execute!

There are examples missing, because documentation calls them.
- connectivity/plot_connect_comparison.py removed in 25bce1f1d9a0a722faa099d95dcce8c478f21b39
called from connectivity/connectome_extraction.rst
I fixed one reference in this file, the other I'm not completely sure which is re replacement file.

This I don't if is related to this PR
- I don't get the user guide for Image manipulation and visualization and Advanced usage: manual pipelines and scaling up. So 2 complete chapters are missing.
- And also I have some missing links appear in manipulating_visualizing/plotting.rst. They are missing in Gael's version too.

**Comment by GaelVaroquaux** [comment]
> There are examples missing, because documentation calls them.

Good catch. Could you add an issue listing all the problems that you have
found in the documentation. We can address it separately.

File: doc/sphinxext/sphinxgallery/__init__.py [hunk 2]
```diff
@@ -0,0 +1,8 @@
+"""Sphinx Gallery
+"""
+import os
+__version__ = '0.0.8'
+
+def path_static():
+    """Returns path to packaged static files"""
+    return os.path.abspath(os.path.dirname(__file__))+'/_static'
```
Comment (GaelVaroquaux) [inline doc/sphinxext/sphinxgallery/__init__.py]: PEP8: you should have spaces before and after the "+" operator. 

In addition, it would be better if you used os.path.join, rather than string concatenation (it avoid hard coding the "/", which varies across OS).

File: doc/sphinxext/sphinxgallery/_static/gallery.css [hunk 3]
```diff
@@ -0,0 +1,62 @@
...
+  background: #FFF;
+  margin: 5px;
+  padding-top: 5px;
+  min-height: 230px;
+  border: solid white 1px;
+  -webkit-border-radius: 5px;
+  -moz-border-radius: 5px;
+  border-radius: 5px;
+  float: left;
+  position: relative; }
```
Comment (GaelVaroquaux) [inline doc/sphinxext/sphinxgallery/_static/gallery.css]: Maybe some white space: "}" on new lines and empty lines between blocks.

File: doc/sphinxext/sphinxgallery/_static/gallery.css [hunk 4]
```diff
@@ -0,0 +1,62 @@
...
+    box-shadow: 0 0 15px rgba(142, 176, 202, 0.5);
+    border: solid #B4DDFC 1px; }
+  div.sphx-glr-thumbContainer a {
+    display: block; }
+  div.sphx-glr-thumbContainer p {
+    margin: 0 0 .1em 0; }
+  div.sphx-glr-thumbContainer .figure {
+    margin: 10px;
+    width: 160px; }
+  div.sphx-glr-thumbContainer img {
```
Comment (GaelVaroquaux) [inline doc/sphinxext/sphinxgallery/_static/gallery.css]: It seems to me that these blocks are not indented correctly.

File: doc/sphinxext/sphinxgallery/gen_gallery.py [hunk 5]
```diff
@@ -0,0 +1,95 @@
...
+===================
+
+""")
+    # Here we don't use an os.walk, but we recurse only twice: flat is
+    # better than nested.
+    seen_backrefs = set()
+    generate_dir_rst('.', fhindex, examples_dir, gallery_dir, gallery_conf, plot_gallery, seen_backrefs)
+    for directory in sorted(os.listdir(examples_dir)):
+        if os.path.isdir(os.path.join(examples_dir, directory)):
+            generate_dir_rst(directory, fhindex, examples_dir, gallery_dir, gallery_conf, plot_gallery, seen_backrefs)
```
Comment (GaelVaroquaux) [inline doc/sphinxext/sphinxgallery/gen_gallery.py]: PEP8: isn't this line too long (more than 79 characters)?

File: doc/sphinxext/sphinxgallery/gen_gallery.py [hunk 6]
```diff
@@ -0,0 +1,95 @@
...
+    #  changes their layout between versions, this will not work (though
+    #  it should probably not cause a crash).  Tested successfully
+    #  on Sphinx 1.0.7
+    build_image_dir = '_build/html/_images'
+    if os.path.exists(build_image_dir):
+        filelist = os.listdir(build_image_dir)
+        for filename in filelist:
+            if filename.endswith('png'):
+                os.remove(os.path.join(build_image_dir, filename))
+
```
Comment (GaelVaroquaux) [inline doc/sphinxext/sphinxgallery/gen_gallery.py]: PEP8: two empty lines should separate top-level definitions.

File: doc/sphinxext/sphinxgallery/gen_rst.py [hunk 7]
```diff
@@ -0,0 +1,603 @@
...
+                                         " example file:\n {}\n and make sure"
+                                         " it's correct".format(filename))
+                else:
+                    first_par = paragraphs[0]
+
+        break
+    return docstring, first_par, erow + 1 + start_row
+
+
+
```
Comment (GaelVaroquaux) [inline doc/sphinxext/sphinxgallery/gen_rst.py]: PEP8: too many empty lines: there should be only 2.

**Comment by GaelVaroquaux** [comment]
Almost there: only a few cosmetic comments. After these are addressed, we are ready to merge.

Congratulations, this is great!

**Comment by lesteve** [comment]
> Almost there: only a few cosmetic comments. After these are addressed, we are ready to merge.

> Congratulations, this is great!

I think your comments are related to sphinx-gallery and not nilearn. I reckon we should merge this PR. The alternative is for your comments to be addressed in sphinx-gallery and a new sphinx-gallery release to be made before we can merge this PR. This seems a little bit too much overhead for what it is worth IMHO.

Just for clarity we'll try to keep the sphinx-gallery copy inside nilearn in sync with the latest sphinx-gallery release going forward, which means that if your comments get addressed in sphinx-gallery they will eventually reach the nilearn sphinx-gallery copy.

**Comment by GaelVaroquaux** [comment]
I agree. Oscar, could you please either address these comments inside sphinx gallery,  or copy them as an issue,  so that they don't get lost.

**Comment by AlexandreAbraham** [comment]
:+1: for merging as-is and address comments in sphinx-gallery repo.

**Comment by lesteve** [comment]
OK merging then, thanks a lot for this and hurray for nilearn officially starting to use sphinx-gallery !
